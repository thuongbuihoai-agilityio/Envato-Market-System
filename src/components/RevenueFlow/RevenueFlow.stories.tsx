import { StoryObj, Meta } from '@storybook/react';

import RevenueFlow from '.';
import { REVENUE_FLOW_MOCK } from '@app/mocks';

const meta: Meta<typeof RevenueFlow> = {
  title: 'Custom Components/RevenueFlow',
  tags: ['autodocs'],
  component: RevenueFlow,
  argTypes: {
    data: {
      description:
        'The data of revenue flow for 12 months includes: pending, signed and lost',
    },
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;
type Story = StoryObj<typeof RevenueFlow>;

export const Default: Story = {
  args: {
    data: REVENUE_FLOW_MOCK,
  },
};
