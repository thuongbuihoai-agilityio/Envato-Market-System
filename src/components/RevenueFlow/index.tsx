import { memo, useCallback, useMemo, useState } from 'react';
import isEqual from 'react-fast-compare';

// Components
import Chart from 'react-apexcharts';
import {
  Box,
  Flex,
  Heading,
  // Skeleton,
  Text,
  theme,
  useColorModeValue,
} from '@chakra-ui/react';
import { Select } from '..';

// Icon
import { Arrow } from '@app/components/Icons';

// Constants
import {
  MONTHS,
  REVENUE_FLOW_COLORS,
  REVENUE_FLOW_OPTIONS,
  REVENUE_FLOW_STATUS,
} from '@app/constants';

// Types
import { IRevenueFlow, RevenueFlowStatus } from '@app/interfaces';
import { TOption } from '@app/components/common/Select';

interface RevenueFlowProps {
  data: IRevenueFlow[];
}

interface ChartDataState {
  data: number[];
}

const RevenueFlowComponent = ({ data }: RevenueFlowProps) => {
  const defaultSeries = useMemo(() => {
    const generatesData = (
      key: keyof Omit<IRevenueFlow, '_id' | 'title'>,
    ): number[] => data.map((item: IRevenueFlow) => item[key]);

    return [
      {
        data: generatesData(RevenueFlowStatus.PENDING),
      },
      {
        data: generatesData(RevenueFlowStatus.SINGED),
      },
      {
        data: generatesData(RevenueFlowStatus.LOST),
      },
    ];
  }, [data]);

  const [chartData, setChartData] = useState<ChartDataState[]>(defaultSeries);

  const colorFill = useColorModeValue(
    theme.colors.gray[800],
    theme.colors.white,
  );

  const renderTitle = useCallback(
    ({ label }: TOption) => (
      <Flex alignItems="center">
        <Text fontSize="sm">{label}</Text>
        <Arrow color={colorFill} />
      </Flex>
    ),
    [colorFill],
  );

  const handleChangeSelect = useCallback(
    (option: TOption) => {
      const [monthStart, monthEnd] = option.value.split(',');

      const rangeMonths = MONTHS.slice(
        MONTHS.findIndex((i) => i === monthStart),
        MONTHS.findIndex((i) => i === monthEnd) + 1,
      );

      const getRevenueFlowDetails = (
        key: keyof Omit<IRevenueFlow, '_id' | 'title'>,
      ) =>
        data.map((revenue: IRevenueFlow) =>
          rangeMonths.includes(revenue.title) ? revenue[key] : 0,
        );
      const result = [
        {
          data: getRevenueFlowDetails(RevenueFlowStatus.PENDING),
        },
        {
          data: getRevenueFlowDetails(RevenueFlowStatus.SINGED),
        },
        {
          data: getRevenueFlowDetails(RevenueFlowStatus.LOST),
        },
        {
          data: data.map((revenue: IRevenueFlow) =>
            rangeMonths.includes(revenue.title)
              ? 0
              : revenue.lost + revenue.pending + revenue.signed,
          ),
        },
      ];

      setChartData(result);
    },
    [data],
  );

  return (
    <Box py={3} px={6} bg="background.component.primary" rounded="lg">
      <Flex
        py={4}
        px={5}
        borderBottom="1px"
        borderColor="border.primary"
        justifyContent="space-between"
      >
        <Heading variant="heading2Xl" as="h3">
          Revenue Flow
        </Heading>
        <Flex gap={7} display={{ base: 'none', lg: 'flex' }}>
          {REVENUE_FLOW_STATUS.map((item, index) => (
            <Flex key={item} gap={2} alignItems="center">
              <Box
                bgColor={REVENUE_FLOW_COLORS[index]}
                w={3}
                height={3}
                rounded="50%"
              />
              <Text variant="textSm">{item}</Text>
            </Flex>
          ))}
        </Flex>
        <Box w={100} h="21px">
          <Select
            options={REVENUE_FLOW_OPTIONS}
            size="sm"
            variant="no-border"
            renderTitle={renderTitle}
            onSelect={handleChangeSelect}
          />
        </Box>
      </Flex>
      <Chart
        options={{
          chart: {
            stacked: true,
            toolbar: {
              show: false,
            },
          },
          xaxis: {
            categories: data.map((item) => item.title),
            axisTicks: {
              show: false,
            },
            labels: {
              style: {
                colors: colorFill,
              },
            },
          },
          yaxis: {
            labels: {
              style: {
                colors: colorFill,
              },
            },
          },
          legend: {
            show: false,
          },
          colors: REVENUE_FLOW_COLORS,
          dataLabels: {
            enabled: false,
          },
          tooltip: {
            custom: function ({ series, seriesIndex, dataPointIndex }) {
              const status = REVENUE_FLOW_STATUS[seriesIndex]
                ? `${REVENUE_FLOW_STATUS[seriesIndex]}:`
                : '';

              return `<div style="padding: 10px; background-color: #000; color: #FFF" >
                <p>
                ${data[dataPointIndex].title}
                </p>
                <span>
                ${status} ${series[seriesIndex][dataPointIndex]}
                </span>
                </div>`;
            },
          },
        }}
        series={chartData}
        type="bar"
        height="230"
      />
    </Box>
  );
};

const RevenueFlow = memo(RevenueFlowComponent, isEqual);

export default RevenueFlow;
