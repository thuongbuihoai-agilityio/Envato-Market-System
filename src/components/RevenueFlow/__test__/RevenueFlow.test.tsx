import { render } from '@testing-library/react';

// Components
import RevenueFlow from '..';

// Mock
import { REVENUE_FLOW_MOCK } from '@mocks/index';

jest.mock('react-apexcharts', () => ({
  __esModule: true,
  default: () => <div />,
}));

describe('RevenueFlow component', () => {
  it('renders correctly', () => {
    const { container } = render(<RevenueFlow data={REVENUE_FLOW_MOCK} />);

    expect(container).toMatchSnapshot();
  });

  it('renders with is loading is true', () => {
    const { container } = render(
      <RevenueFlow data={REVENUE_FLOW_MOCK} isLoading />,
    );

    expect(container).toMatchSnapshot();
  });
});
