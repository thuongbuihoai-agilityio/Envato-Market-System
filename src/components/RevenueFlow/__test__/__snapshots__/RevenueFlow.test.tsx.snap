// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RevenueFlow component renders correctly 1`] = `
<div>
  <div
    class="css-hfln1x"
  >
    <div
      class="css-1spmqp"
    >
      <h3
        class="chakra-heading css-0"
      >
        Revenue Flow
      </h3>
      <div
        class="css-vd6ngq"
      >
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-18ek59d"
          />
          <p
            class="chakra-text css-0"
          >
            Pending
          </p>
        </div>
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-vdd5pp"
          />
          <p
            class="chakra-text css-0"
          >
            Signed
          </p>
        </div>
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-y7nxfq"
          />
          <p
            class="chakra-text css-0"
          >
            Lost
          </p>
        </div>
      </div>
      <div
        class="css-19vtkro"
      >
        <button
          aria-controls="menu-list-:r1:"
          aria-expanded="false"
          aria-haspopup="menu"
          class="chakra-button chakra-menu__menu-button css-9v42w8"
          id="menu-button-:r1:"
          type="button"
        >
          <span
            class="css-xl71ch"
          >
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-0"
              >
                Jan - Dec
              </p>
              <svg
                fill="none"
                height="17"
                stroke="#1A202C"
                viewBox="0 0 17 17"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 10L12 14L17 10"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </div>
          </span>
        </button>
        <div
          class="css-0"
          style="visibility: hidden; position: absolute; inset: 0 auto auto 0;"
        >
          <div
            aria-orientation="vertical"
            class="chakra-menu__menu-list css-ta0hp9"
            id="menu-list-:r1:"
            role="menu"
            style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
            tabindex="-1"
          >
            <button
              class="chakra-menu__menuitem css-1jeunnm"
              data-index="0"
              id="menu-list-:r1:-menuitem-:r2:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                Jan - Dec
              </p>
            </button>
            <button
              class="chakra-menu__menuitem css-1jeunnm"
              data-index="1"
              id="menu-list-:r1:-menuitem-:r3:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                Jan - Jun
              </p>
            </button>
            <button
              class="chakra-menu__menuitem css-1jeunnm"
              data-index="2"
              id="menu-list-:r1:-menuitem-:r4:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                July - Dec
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</div>
`;

exports[`RevenueFlow component renders with is loading is true 1`] = `
<div>
  <div
    class="css-hfln1x"
  >
    <div
      class="css-1spmqp"
    >
      <h3
        class="chakra-heading css-0"
      >
        Revenue Flow
      </h3>
      <div
        class="css-vd6ngq"
      >
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-18ek59d"
          />
          <p
            class="chakra-text css-0"
          >
            Pending
          </p>
        </div>
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-vdd5pp"
          />
          <p
            class="chakra-text css-0"
          >
            Signed
          </p>
        </div>
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-y7nxfq"
          />
          <p
            class="chakra-text css-0"
          >
            Lost
          </p>
        </div>
      </div>
      <div
        class="css-19vtkro"
      >
        <button
          aria-controls="menu-list-:r6:"
          aria-expanded="false"
          aria-haspopup="menu"
          class="chakra-button chakra-menu__menu-button css-9v42w8"
          id="menu-button-:r6:"
          type="button"
        >
          <span
            class="css-xl71ch"
          >
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-0"
              >
                Jan - Dec
              </p>
              <svg
                fill="none"
                height="17"
                stroke="#1A202C"
                viewBox="0 0 17 17"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 10L12 14L17 10"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </div>
          </span>
        </button>
        <div
          class="css-0"
          style="visibility: hidden; position: absolute; inset: 0 auto auto 0;"
        >
          <div
            aria-orientation="vertical"
            class="chakra-menu__menu-list css-ta0hp9"
            id="menu-list-:r6:"
            role="menu"
            style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
            tabindex="-1"
          >
            <button
              class="chakra-menu__menuitem css-1jeunnm"
              data-index="0"
              id="menu-list-:r6:-menuitem-:r7:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                Jan - Dec
              </p>
            </button>
            <button
              class="chakra-menu__menuitem css-1jeunnm"
              data-index="1"
              id="menu-list-:r6:-menuitem-:r8:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                Jan - Jun
              </p>
            </button>
            <button
              class="chakra-menu__menuitem css-1jeunnm"
              data-index="2"
              id="menu-list-:r6:-menuitem-:r9:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                July - Dec
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-skeleton css-1x2ptk6"
    />
  </div>
</div>
`;
